### PR TITLE
fix: `.keep` in `mutate()` doesn't keep columns correctly

### LINKED ISSUE #353: `mutate(..., .keep = c("all"/"used"/"unnused"/"none"))` doesn't keep columns correctly
**Describe the bug**

There are problems with the current `mutate(..., .keep = c("all"/"used"/"unnused"/"none"))` argument. Please see the below reproduce code for details:

**To Reproduce**

``` r
library(dplyr)
library(tidypolars)

df <- tibble(
  x = c(1, 2, 3),
  y = c(4, 5, 6)
)
df_pl <- as_polars_df(df)

# Problem:
# If a column is overwritten but not used in expression,
# the modified column will be deleted.
df |> mutate(x = y + 1, .keep = "used")
#> # A tibble: 3 × 2
#>       x     y
#>   <dbl> <dbl>
#> 1     5     4
#> 2     6     5
#> 3     7     6
df_pl |> mutate(x = y + 1, .keep = "used")
#> shape: (3, 1)
#> ┌─────┐
#> │ y   │
#> │ --- │
#> │ f64 │
#> ╞═════╡
#> │ 4.0 │
#> │ 5.0 │
#> │ 6.0 │
#> └─────┘

# Problem:
# If a column is overwritten and used in expression,
# the modified column will be deleted.
df |> mutate(x = x + 1, .keep = "unused")
#> # A tibble: 3 × 2
#>       x     y
#>   <dbl> <dbl>
#> 1     2     4
#> 2     3     5
#> 3     4     6
df_pl |> mutate(x = x + 1, .keep = "unused")
#> shape: (3, 1)
#> ┌─────┐
#> │ y   │
#> │ --- │
#> │ f64 │
#> ╞═════╡
#> │ 4.0 │
#> │ 5.0 │
#> │ 6.0 │
#> └─────┘

# Problem:
# If a column is overwritten,
# the modified column will be deleted.
df |> mutate(x = x + 1, .keep = "none")
#> # A tibble: 3 × 1
#>       x
#>   <dbl>
#> 1     2
#> 2     3
#> 3     4
df_pl |> mutate(x = x + 1, .keep = "none")
#> shape: (0, 0)
#> ┌┐
#> ╞╡
#> └┘

# Problem:
# After a column is deleted by `NULL`, `.keep` may delete it again.
df |> mutate(x = NULL, .keep = "unused")
#> # A tibble: 3 × 1
#>       y
#>   <dbl>
#> 1     4
#> 2     5
#> 3     6
df_pl |> mutate(x = NULL, .keep = "unused")
#> Error in `.data$drop()`:
#> ! Evaluation failed in `$drop()`.
#> Caused by error:
#> ! Evaluation failed in `$drop()`.
#> Caused by error:
#> ! `...` can only contain single strings or polars selectors.

# Problem:
# When use `group_by()` and then modify the group column,
# the modified group column will be deleted, then be used to group again.
df |>
  group_by(x) |>
  mutate(x = x + 1, .keep = "unused")
#> # A tibble: 3 × 2
#> # Groups:   x [3]
#>       x     y
#>   <dbl> <dbl>
#> 1     2     4
#> 2     3     5
#> 3     4     6
df_pl |>
  group_by(x) |>
  mutate(x = x + 1, .keep = "unused")
#> Error in `group_by()`:
#> ℹ In argument: `all_of(grps)`.
#> Caused by error in `all_of()`:
#> ! Can't subset elements that don't exist.
#> ✖ Element `x` doesn't exist.

# Problem:
# When use `.by`, the modified group column will be deleted.
df |>
  mutate(x = x + 1, .by = x, .keep = "unused")
#> # A tibble: 3 × 2
#>       x     y
#>   <dbl> <dbl>
#> 1     2     4
#> 2     3     5
#> 3     4     6
df_pl |>
  mutate(x = x + 1, .by = x, .keep = "unused")
#> shape: (3, 1)
#> ┌─────┐
#> │ y   │
#> │ --- │
#> │ f64 │
#> ╞═════╡
#> │ 4.0 │
#> │ 5.0 │
#> │ 6.0 │
#> └─────┘
```

**Expected behavior**

Same as `dplyr`.

I'll open a PR.

### PR DESCRIPTION
**Pull request by Yousa-Mirage**
fix #353

### CODE REVIEW

File: NEWS.md [hunk 1]
```diff
@@ -4,6 +4,10 @@
 
 * Added support for `%notin%` (#349).
 
+## Bug fixes
+
+- Fix `.keep` in `mutate()` to avoid accidentally droping columns (@Yousa-Mirage, #353).
```
Comment (etiennebacher) [inline NEWS.md]:
```suggestion
- Fix `.keep` in `mutate()` to avoid accidentally dropping columns (@Yousa-Mirage, #353).
```

The wordlist file shouldn't be modified.

File: tests/testthat/test-mutate.R [hunk 2]
```diff
@@ -386,6 +386,84 @@ test_that("argument .keep works", {
   )
 })
 
+test_that("argument .keep preserves overwritten columns", {
```
Comment (etiennebacher) [inline tests/testthat/test-mutate.R:389]: Can we parametrize this and the other new tests with `patrick`? There are a few examples in other tests.

**Review by etiennebacher — CHANGES_REQUESTED** [review]
Thanks! A couple of comments to address

File: NEWS.md (hunk 1 above)
Comment (Yousa-Mirage) [reply NEWS.md]: Oh yeah. My English is bad :(

Comment (etiennebacher) [reply NEWS.md]: No worries, I'm pretty sure the docs contain mistakes and sentence structures that don't work very well in English.

File: tests/testthat/test-mutate.R (hunk 2 above)
Comment (Yousa-Mirage) [reply tests/testthat/test-mutate.R:389]: This might not be a good idea. Because different parameters will only raise errors under different specific expressions, it is difficult to pass parameters in batches through `patrick`. Currently I selected the edge cases that raise errors before as the test cases.

**Comment by Yousa-Mirage** [comment]
These three recent PRs are all modifications to `mutate()`, and may have some overlapping logic. Sorry for opening so many PRs at once, as I finally have some free time and energy to maintain them today. I would recommend you review this PR first as it is the simplest to implement. Then maybe #357 for the `.before` and `.after` parameters, since the modification is mainly about calling existing functions.
